aggregator/aggregator.go

```diff
@@ -428,14 +427,45 @@ func (agg *aggregator) checkMetricType(mu unaggregated.MetricUnion) error {
 	}
 }
 
-func (agg *aggregator) ownedShards() []*aggregatorShard {
-	agg.RLock()
-	ownedShards := make([]*aggregatorShard, len(agg.shardIDs))
-	for i, shardID := range agg.shardIDs {
-		ownedShards[i] = agg.shards[shardID]
+func (agg *aggregator) ownedShards() (owned, toClose []*aggregatorShard) {
+	agg.Lock()
+	defer agg.Unlock()
+
+	owned = make([]*aggregatorShard, 0, len(agg.shardIDs))
+	for i := 0; i < len(agg.shardIDs); i++ {
+		shardID := agg.shardIDs[i]
+		// NB(xichen): if traffic has been cut off for a shard, the aggregator actively
+		// closes it to avoid wasting CPU cycles flushing its metric lists.
+		if shard := agg.shards[shardID]; !shard.IsCutoff() {
+			owned = append(owned, shard)
+		} else {
+			lastIdx := len(agg.shardIDs) - 1
+			agg.shardIDs[i], agg.shardIDs[lastIdx] = agg.shardIDs[lastIdx], agg.shardIDs[i]
+			agg.shardIDs = agg.shardIDs[:lastIdx]
+			i--
+			agg.shards[shardID] = nil
+			toClose = append(toClose, shard)
+		}
+	}
+	return owned, toClose
+}
+
+// closeShardsAsync asynchronously closes the shards to avoid blocking writes.
+// Because each shard write happens while holding the shard read lock, the shard
+// may only close itself after all its pending writes are finished.
+func (agg *aggregator) closeShardsAsync(shards []*aggregatorShard) {
+	atomic.AddInt32(&agg.shardsPendingClose, int32(len(shards)))
+	agg.metrics.shards.pendingClose.Update(float64(atomic.LoadInt32(&agg.shardsPendingClose)))
+
+	for _, shard := range shards {
+		shard := shard
+		go func() {
+			shard.Close()
+			atomic.AddInt32(&agg.shardsPendingClose, -1)
+			agg.metrics.shards.pendingClose.Update(float64(atomic.LoadInt32(&agg.shardsPendingClose)))
```

Comment:
nit: `atomic.AddInt32` returns the new value of the integer so we don't need the additional load here.

Reply:
Sounds good.